pkg/certifier/osv/osv.go

```diff
@@ -0,0 +1,196 @@
+//
+// Copyright 2022 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package osv
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/guacsec/guac/pkg/assembler"
+	"github.com/guacsec/guac/pkg/certifier"
+	attestation_vuln "github.com/guacsec/guac/pkg/certifier/attestation"
+	"github.com/guacsec/guac/pkg/certifier/osv/internal/osv_query"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+	osv_scanner "golang.org/x/vuln/osv"
+)
+
+const (
+	URI         string = "osv.dev"
+	VERSION     string = "0.0.14"
+	INVOC_URI   string = "guac"
+	PRODUCER_ID string = "guecsec/guac"
+)
+
+type osvCertifier struct {
+	rootComponents *certifier.Component
+}
+
+// NewOSVCertificationParser initializes the OSVCertifier
+func NewOSVCertificationParser() certifier.Certifier {
+	return &osvCertifier{}
+}
+
+// CertifyVulns takes in the root component from the gauc database and does a recursive scan
+// to generate vulnerability attestations
+func (o *osvCertifier) CertifyVulns(ctx context.Context, rootComponent *certifier.Component, docChannel chan<- *processor.Document) error {
+	o.rootComponents = rootComponent
+	_, err := o.certifyHelper(ctx, rootComponent, rootComponent.DepPackages, docChannel)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *osvCertifier) certifyHelper(ctx context.Context, topLevel *certifier.Component, depPackages []*certifier.Component, docChannel chan<- *processor.Document) ([]osv_scanner.Entry, error) {
+	packNodes := []assembler.PackageNode{}
+	totalDepVul := []osv_scanner.Entry{}
+	for _, depPack := range depPackages {
+		if len(depPack.DepPackages) > 0 {
+			depVulns, err := o.certifyHelper(ctx, depPack, depPack.DepPackages, docChannel)
+			if err != nil {
+				return nil, nil
+			}
+			totalDepVul = append(totalDepVul, depVulns...)
+		} else {
+			packNodes = append(packNodes, depPack.CurPackage)
```

Comment:
hmm, would refactoring this to recursively return the entries for a component streamline the function a bit more? I found it a bit confusing to follow.. or unless i'm looking at it a different way, maybe a comment to help out what's going on.

Reply:
Added the following comment:
```
// certifyHelper recursively checks each component for dependencies. If it has dependencies, certifyHelper is re-called until no more dependencies are found.
// The dependency node is appended to the package node array and sent to be queried by OSV
// Once the vulnerabilities are found, an attestation is generated for that package node
// All the vulnerabilities for each dependent package node are collected and the parent package node's
// attestation is generated containing all the vulnerabilities of its dependencies
// these vulnerabilities are passed up until it reaches the root level node which contains an attestation
// with all the aggregate vulnerabilities
```